clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,268 @@
+                    netInBuffer = null; // reset the networkReceive as we read all the data.
+                    sendSaslToken(serverToken);
+                }
+                if (saslClient.isComplete())
+                    saslState = SaslState.COMPLETE;
+                break;
+            case COMPLETE:
+                break;
+            case FAILED:
+                throw new IOException("SASL handshake failed");
+        }
+    }
+
+    private void sendSaslToken(byte[] serverToken) throws IOException {
+        if (!saslClient.isComplete()) {
+            try {
+                saslToken = createSaslToken(serverToken);
+                if (saslToken != null) {
+                    netOutBuffer = new NetworkSend(node, ByteBuffer.wrap(saslToken));
+                    if (!flushNetOutBuffer())
+                        transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+                }
+            } catch (SaslException se) {
+                saslState = SaslState.FAILED;
+                throw new IOException("Failed to authenticate using SASL " + se);
+            }
+        }
+    }
+
+    public Principal principal() {
+        return KafkaPrincipal.ANONYMOUS;
+    }
+
+    public boolean complete() {
+        return saslClient.isComplete() && saslState == SaslState.COMPLETE;
+    }
+
+    public void close() throws IOException {
+        saslClient.dispose();
+    }
+
+    private byte[] createSaslToken(final byte[] saslToken) throws SaslException {
+        if (saslToken == null) {
+            throw new SaslException("Error in authenticating with a Kafka Broker: the kafka broker saslToken is null.");
+        }
+
+        try {
+            final byte[] retval =
+                Subject.doAs(subject, new PrivilegedExceptionAction<byte[]>() {
+                        public byte[] run() throws SaslException {
+                            return saslClient.evaluateChallenge(saslToken);
+                        }
+                    });
+            return retval;
+        } catch (PrivilegedActionException e) {
+            String error = "An error: (" + e + ") occurred when evaluating Kafka Brokers " +
+                      " received SASL token.";
+            // Try to provide hints to use about what went wrong so they can fix their configuration.
+            // TODO: introspect about e: look for GSS information.
+            final String unknownServerErrorText =
+                "(Mechanism level: Server not found in Kerberos database (7) - UNKNOWN_SERVER)";
+            if (e.toString().indexOf(unknownServerErrorText) > -1) {
+                error += " This may be caused by Java's being unable to resolve the Kafka Broker's" +
+                    " hostname correctly. You may want to try to adding" +
+                    " '-Dsun.net.spi.nameservice.provider.1=dns,sun' to your client's JVMFLAGS environment." +
+                    " Users must configure FQDN of kafka brokers when authenticating using SASL.";
```

Comment:
Let's also add "socketChannel.socket().getInetAddress().getHostName() must match the hostname in principal/hostname@realm"